drivers/adc/adc_ads114s08.c

```diff
@@ -0,0 +1,969 @@
+		/* enable positive reference buffer */
+		ADS114_REGISTER_REF_NOT_REFP_BUF_SET(reference_control, 0b0);
+		/* use external reference 0*/
+		ADS114_REGISTER_REF_REFSEL_SET(reference_control, 0b01);
+		break;
+	default:
+		LOG_ERR("reference %i is not supported", channel_cfg->reference);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->differential) {
+		if (channel_cfg->input_positive >= ADS114_INPUT_SELECTION_AINCOM) {
+			LOG_ERR("positive channel input %i is invalid",
+				channel_cfg->input_positive);
+			return -EINVAL;
+		}
+
+		if (channel_cfg->input_negative >= ADS114_INPUT_SELECTION_AINCOM) {
+			LOG_ERR("negative channel input %i is invalid",
+				channel_cfg->input_positive);
+			return -EINVAL;
+		}
+
+		if (channel_cfg->input_positive == channel_cfg->input_negative) {
+			LOG_ERR("negative and positive channel inputs must be different");
+			return -EINVAL;
+		}
+
+		ADS114_REGISTER_INPMUX_MUXP_SET(input_mux, channel_cfg->input_positive);
+		ADS114_REGISTER_INPMUX_MUXN_SET(input_mux, channel_cfg->input_negative);
+	} else {
+		if (channel_cfg->channel_id >= ADS114_INPUT_SELECTION_AINCOM) {
+			LOG_ERR("channel input %i is invalid", channel_cfg->input_positive);
+			return -EINVAL;
+		}
+
+		ADS114_REGISTER_INPMUX_MUXP_SET(input_mux, channel_cfg->input_positive);
+		ADS114_REGISTER_INPMUX_MUXN_SET(input_mux, ADS114_INPUT_SELECTION_AINCOM);
+	}
+
+	switch (channel_cfg->gain) {
+	case ADC_GAIN_1:
+		/* set gain value */
+		ADS114_REGISTER_PGA_GAIN_SET(gain, 0b000);
+		break;
+	case ADC_GAIN_2:
+		ADS114_REGISTER_PGA_GAIN_SET(gain, 0b001);
+		break;
+	case ADC_GAIN_4:
+		ADS114_REGISTER_PGA_GAIN_SET(gain, 0b010);
+		break;
+	case ADC_GAIN_8:
+		ADS114_REGISTER_PGA_GAIN_SET(gain, 0b011);
+		break;
+	case ADC_GAIN_16:
+		ADS114_REGISTER_PGA_GAIN_SET(gain, 0b100);
+		break;
+	case ADC_GAIN_32:
+		ADS114_REGISTER_PGA_GAIN_SET(gain, 0b101);
+		break;
+	case ADC_GAIN_64:
+		ADS114_REGISTER_PGA_GAIN_SET(gain, 0b110);
+		break;
+	case ADC_GAIN_128:
+		ADS114_REGISTER_PGA_GAIN_SET(gain, 0b111);
+		break;
+	default:
+		LOG_ERR("gain value %i not supported", channel_cfg->gain);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		/* enable gain */
+		ADS114_REGISTER_PGA_PGA_EN_SET(gain, 0b01);
+	}
+
+	result = ads114_write_register(&config->bus, ADS114_REGISTER_INPMUX, input_mux);
+
+	if (result != 0) {
+		LOG_ERR("unable to configure input mux");
+		return result;
+	}
+
+	result = ads114_write_register(&config->bus, ADS114_REGISTER_REF, reference_control);
+
+	if (result != 0) {
+		LOG_ERR("unable to configure reference control");
+		return result;
+	}
+
+	result = ads114_write_register(&config->bus, ADS114_REGISTER_DATARATE, data_rate);
+
+	if (result != 0) {
+		LOG_ERR("unable to configure data rate");
+		return result;
+	}
+
+	result = ads114_write_register(&config->bus, ADS114_REGISTER_PGA, gain);
+
+	if (result != 0) {
+		LOG_ERR("unable to configure PGA");
+		return result;
+	}
```

Comment:
These writes can be combined into a single multi-register write by writing these registers in order (INPMUX, PGA, DATARATE, REF) by extending ads114_write_register() to support a length parameter.

Refer to [lmp90xxx_adc_channel_setup()](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/adc/adc_lmp90xxx.c#L458) for an example.

Reply:
Good idea, I just modified the approach slightly to ensure that the caller has an assert check that the correct registers are written to.